volumina/tiling/cache.py

```diff
@@ -0,0 +1,299 @@
+import threading
+import warnings
+import logging
+from collections import OrderedDict, defaultdict
+
+import numpy
+from PyQt5.QtWidgets import QGraphicsItem
+
+logger = logging.getLogger()
+
+__all__ = ["TilesCache"]
+
+
+class CachePolicy:
+    def __init__(self, size):
+        self._validate_size(size)
+        self._size = size
+        self._subsribers = []
+
+    @property
+    def size(self):
+        return self._size
+
+    def _validate_size(self, value):
+        if not isinstance(value, int) or value <= 0:
+            raise ValueError("size should non negative integer")
+
+    def set_size(self, value):
+        if self._size == value:
+            return
+
+        self._validate_size(value)
+
+        self._size = value
+
+        for sub in self._subsribers:
+            sub()
+
+    def subscribe(self, fn):
+        self._subsribers.append(fn)
+
+    def unsubsribe(self, fn):
+        try:
+            self._subsribers.remove(fn)
+        except ValueError:
+            pass
+
+
+class MultiCacheException(Exception):
+    pass
+
+
+class MultiCache:
```

Comment:
`abc.Sequence` expects `__getitem__` method to be indexable, not keyable. Running tests fails with:
```    
def __getitem__(self, key):
>       return self._caches[key]
E       KeyError: 0
```

Reply:
What about `Mapping` then?

Reply:
The question is what is the benefit of using `Mapping` interface here.
On the other hand, if I were to use `Mapping` it can be confusing because when I think about `Mapping` I expect a hashmap with certain behavior, which should not throw away keys.

Reply:
`Mapping` does not guarantee hashmap-like behaviour, only existence of a certain methods. For example, cache objects in [cachetools](https://github.com/tkem/cachetools/) are mappings.

Reply:
My point is mostly if I had a class where I used `Mapping` which can be substituted for either for `MultiCache` or some other `Mapping` I would then inherit from this abstract class. 
But the only place where we use `MultiCache` excepts `MultiCache` and not a `Mapping` so inheriting mapping doesn't really provide any benefit in terms of decoupling.